agent/node.go

```diff
@@ -240,28 +240,31 @@ func (n *Node) run(ctx context.Context) (err error) {
 		return err
 	}
 
-	if n.role == ca.ManagerRole {
-		n.managerRoleCh <- struct{}{}
-	}
-
 	forceCertRenewal := make(chan struct{})
 	go func() {
-		n.RLock()
-		lastRole := n.role
-		n.RUnlock()
 		for {
 			select {
 			case <-ctx.Done():
 				return
 			case apirole := <-n.roleChangeReq:
+				n.RLock()
+				lastRole := n.role
+				n.RUnlock()
 				role := ca.AgentRole
 				if apirole == api.NodeRoleManager {
 					role = ca.ManagerRole
 				}
-				if lastRole != role {
-					forceCertRenewal <- struct{}{}
+				if lastRole == role {
+					continue
+				}
+				n.Lock()
+				// switch role to agent immediately to shutdown manager early
+				if role == ca.AgentRole {
+					n.role = role
+					n.roleCond.Broadcast()
 				}
-				lastRole = role
+				n.Unlock()
+				forceCertRenewal <- struct{}{}
```

Comment:
I wonder if there's also a risk of deadlocking here. `RenewTLSConfig` only returns on `<-ctx.Done()`, but if the context gets cancelled right here, `RenewTLSConfig` might return before this statement is reached.